theano/gpuarray/opt.py

```diff
@@ -895,7 +903,34 @@ def local_gpua_subtensor(op, context_name, inputs, outputs):
                                               for v in n.inputs + n.outputs])
                         for n, _ in clients[0]]):
                     return
+    # Here is the condition for the GraphToGPU opt. inputs is the
+    # inputs we want to use for the new node
+    if (x.owner and isinstance(x.owner.op, GpuFromHost)):
+        cpu_x = x.owner.inputs[0]
+        # And it is a shared var or an input of the graph.
+        # and is used by only 1 node.
+        # x is in the new graph, so we can't tests its number of clients.
+        if not cpu_x.owner and len(cpu_x.clients) == 1:
+            return
+    return GpuSubtensor(op.idx_list)
+
 
+@register_opt2([tensor.Subtensor], 'fast_compile')
+def local_gpua_subtensor_graph(op, context_name, inputs, outputs):
+    # We need different code as the condition is different as inputs
+    # aren't the same.
```

Comment:
in the old opt, inputs[0] is a TensorVariable object. When it is the corresponding of an input of the graph the pattern is HostFromGpu(GpuFromHost(x))

In the new opt, inputs[0] is an GpuArrayVariable. When it correspond to the input of the graph, it is HostFromGpu(x).